central/processlisteningonport/store/postgres/store.go

```diff
@@ -0,0 +1,763 @@
+// Code originally generated by pg-bindings generator.
+
+// XXX: There are custom bits applied on top of the generated code. Everything
+// what comes together with GetProcessListeningOnPort and
+// GetProcessListeningOnPortByNamespace is manually written.
+
+package postgres
+
+import (
+	"context"
+	"time"
+
+	"github.com/gogo/protobuf/proto"
+	"github.com/hashicorp/go-multierror"
+	"github.com/jackc/pgx/v4"
+	"github.com/jackc/pgx/v4/pgxpool"
+	"github.com/pkg/errors"
+	"github.com/stackrox/rox/central/metrics"
+	"github.com/stackrox/rox/central/role/resources"
+	v1 "github.com/stackrox/rox/generated/api/v1"
+	"github.com/stackrox/rox/generated/storage"
+	"github.com/stackrox/rox/pkg/auth/permissions"
+	"github.com/stackrox/rox/pkg/logging"
+	ops "github.com/stackrox/rox/pkg/metrics"
+	"github.com/stackrox/rox/pkg/postgres/pgutils"
+	pkgSchema "github.com/stackrox/rox/pkg/postgres/schema"
+	"github.com/stackrox/rox/pkg/sac"
+	"github.com/stackrox/rox/pkg/search"
+	"github.com/stackrox/rox/pkg/search/postgres"
+	"github.com/stackrox/rox/pkg/sync"
+	"gorm.io/gorm"
+)
+
+const (
+	baseTable = "process_listening_on_ports"
+
+	batchAfter = 100
+
+	// using copyFrom, we may not even want to batch.  It would probably be simpler
+	// to deal with failures if we just sent it all.  Something to think about as we
+	// proceed and move into more e2e and larger performance testing
+	batchSize = 10000
+
+	cursorBatchSize = 50
+	deleteBatchSize = 5000
+
+	// SQL query to join process_listening_on_port together with
+	// process_indicators. Used to provide information for queries like 'give
+	// me all PLOP by this deployment'.
+	// XXX: Verify the query plan to make sure needed indexes are in use.
+	getByDeploymentStmt = "SELECT plop.id, plop.serialized, " +
+		"proc.podid, proc.containername, " +
+		"proc.signal_containerid, " +
+		"proc.signal_name, proc.signal_args, proc.signal_execfilepath, " +
+		"proc.clusterid, proc.serialized as procSerialized " +
+		"FROM process_listening_on_ports plop " +
+		"JOIN process_indicators proc " +
+		"ON plop.processindicatorid = proc.id " +
+		"WHERE proc.deploymentid = $1 AND plop.closed = false"
```

Comment:
AFAIK, In the current state of things, the pattern applied in our app is to duplicate the data from the serialized object in the table columns that can be used for searching.
As a consequence, I think the query here could simply select `plop.id, plop.serialized, proc.serialized as procSerialzed`. This would slightly reduce the amount of data sent over the wire, and decoded by the process.

Reply:
It's probably fine to use this approach for this particular PR, if everything else works this way. But in long term perspective such duplication is a huge problem (more data to store, bulletproof mechanism is required to make sure columns are really the same as in serialized field) and has to be addressed.